// @theia/process/lib/node/process-manager.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * Copyright (C) 2017 Ericsson and others.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 */
var inversify_1 = require("inversify");
var common_1 = require("@theia/core/lib/common");
var logger_1 = require("@theia/core/lib/common/logger");
var ProcessManager = /** @class */ (function () {
    function ProcessManager(logger) {
        this.logger = logger;
        this.id = 0;
        this.processes = new Map();
        this.deleteEmitter = new common_1.Emitter();
    }
    /**
     * Registers the given process into this manager. Both on process termination and on error,
     * the process will be automatically removed from the manager.
     *
     * @param process the process to register.
     */
    ProcessManager.prototype.register = function (process) {
        var _this = this;
        var id = this.id;
        this.processes.set(id, process);
        process.onExit(function () { return _this.unregister(process); });
        process.onError(function () { return _this.unregister(process); });
        this.id++;
        return id;
    };
    /**
     * Removes the process from this process manager. Invoking this method, will make
     * sure that the process is terminated before eliminating it from the manager's cache.
     *
     * @param process the process to unregister from this process manager.
     */
    ProcessManager.prototype.unregister = function (process) {
        var processLabel = this.getProcessLabel(process);
        this.logger.debug("Unregistering process. " + processLabel);
        if (!process.killed) {
            this.logger.debug("Ensuring process termination. " + processLabel);
            process.kill();
        }
        if (this.processes.delete(process.id)) {
            this.deleteEmitter.fire(process.id);
            this.logger.debug("The process was successfully unregistered. " + processLabel);
        }
        else {
            this.logger.warn("This process was not registered or was already unregistered. " + processLabel);
        }
    };
    ProcessManager.prototype.get = function (id) {
        return this.processes.get(id);
    };
    Object.defineProperty(ProcessManager.prototype, "onDelete", {
        get: function () {
            return this.deleteEmitter.event;
        },
        enumerable: true,
        configurable: true
    });
    ProcessManager.prototype.onStop = function () {
        try {
            for (var _a = __values(this.processes.values()), _b = _a.next(); !_b.done; _b = _a.next()) {
                var process_1 = _b.value;
                try {
                    this.unregister(process_1);
                }
                catch (error) {
                    this.logger.error("Error occurred when unregistering process. " + this.getProcessLabel(process_1), error);
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
            }
            finally { if (e_1) throw e_1.error; }
        }
        var e_1, _c;
    };
    ProcessManager.prototype.getProcessLabel = function (process) {
        return "[ID: " + process.id + "]";
    };
    ProcessManager = __decorate([
        inversify_1.injectable(),
        __param(0, inversify_1.inject(logger_1.ILogger)),
        __metadata("design:paramtypes", [Object])
    ], ProcessManager);
    return ProcessManager;
}());
exports.ProcessManager = ProcessManager;
