// @theia/core/lib/browser/common-frontend-contribution.js
"use strict";
/*
 * Copyright (C) 2017 TypeFox and others.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var menu_1 = require("../common/menu");
var message_service_1 = require("../common/message-service");
var application_shell_1 = require("./shell/application-shell");
var tab_bars_1 = require("./shell/tab-bars");
var browser = require("./browser");
var CommonMenus;
(function (CommonMenus) {
    CommonMenus.FILE = __spread(menu_1.MAIN_MENU_BAR, ['1_file']);
    CommonMenus.FILE_NEW = __spread(CommonMenus.FILE, ['1_new']);
    CommonMenus.FILE_OPEN = __spread(CommonMenus.FILE, ['2_open']);
    CommonMenus.FILE_SAVE = __spread(CommonMenus.FILE, ['3_save']);
    CommonMenus.FILE_CLOSE = __spread(CommonMenus.FILE, ['4_close']);
    CommonMenus.EDIT = __spread(menu_1.MAIN_MENU_BAR, ['2_edit']);
    CommonMenus.EDIT_UNDO = __spread(CommonMenus.EDIT, ['1_undo']);
    CommonMenus.EDIT_CLIPBOARD = __spread(CommonMenus.EDIT, ['2_clipboard']);
    CommonMenus.EDIT_FIND = __spread(CommonMenus.EDIT, ['3_find']);
    CommonMenus.VIEW = __spread(menu_1.MAIN_MENU_BAR, ['3_view']);
    CommonMenus.VIEW_VIEWS = __spread(CommonMenus.VIEW, ['1_views']);
    CommonMenus.VIEW_LAYOUT = __spread(CommonMenus.VIEW, ['2_layout']);
    CommonMenus.HELP = __spread(menu_1.MAIN_MENU_BAR, ["4_help"]);
})(CommonMenus = exports.CommonMenus || (exports.CommonMenus = {}));
var CommonCommands;
(function (CommonCommands) {
    CommonCommands.CUT = {
        id: 'core.cut',
        label: 'Cut'
    };
    CommonCommands.COPY = {
        id: 'core.copy',
        label: 'Copy'
    };
    CommonCommands.PASTE = {
        id: 'core.paste',
        label: 'Paste'
    };
    CommonCommands.UNDO = {
        id: 'core.undo',
        label: 'Undo'
    };
    CommonCommands.REDO = {
        id: 'core.redo',
        label: 'Redo'
    };
    CommonCommands.FIND = {
        id: 'core.find',
        label: 'Find'
    };
    CommonCommands.REPLACE = {
        id: 'core.replace',
        label: 'Replace'
    };
    CommonCommands.NEXT_TAB = {
        id: 'core.nextTab',
        label: 'Switch to Next Tab'
    };
    CommonCommands.PREVIOUS_TAB = {
        id: 'core.previousTab',
        label: 'Switch to Previous Tab'
    };
    CommonCommands.CLOSE_TAB = {
        id: 'core.close.tab',
        label: 'Close Tab'
    };
    CommonCommands.CLOSE_OTHER_TABS = {
        id: 'core.close.other.tabs',
        label: 'Close Other Tabs'
    };
    CommonCommands.CLOSE_RIGHT_TABS = {
        id: 'core.close.right.tabs',
        label: 'Close Tabs to the Right'
    };
    CommonCommands.CLOSE_ALL_TABS = {
        id: 'core.close.all.tabs',
        label: 'Close All Tabs'
    };
    CommonCommands.COLLAPSE_PANEL = {
        id: 'core.collapse.tab',
        label: 'Collapse Side Panel'
    };
    CommonCommands.COLLAPSE_ALL_PANELS = {
        id: 'core.collapse.all.tabs',
        label: 'Collapse All Side Panels'
    };
    CommonCommands.TOGGLE_BOTTOM_PANEL = {
        id: 'core.toggle.bottom.panel',
        label: 'Toggle Bottom Panel'
    };
    CommonCommands.SAVE = {
        id: 'core.save',
        label: 'Save'
    };
    CommonCommands.SAVE_ALL = {
        id: 'core.saveAll',
        label: 'Save All'
    };
    CommonCommands.QUIT = {
        id: 'core.quit',
        label: 'Quit'
    };
})(CommonCommands = exports.CommonCommands || (exports.CommonCommands = {}));
exports.supportCut = browser.isNative || document.queryCommandSupported('cut');
exports.supportCopy = browser.isNative || document.queryCommandSupported('copy');
// Chrome incorrectly returns true for document.queryCommandSupported('paste')
// when the paste feature is available but the calling script has insufficient
// privileges to actually perform the action
exports.supportPaste = browser.isNative || (!browser.isChrome && document.queryCommandSupported('paste'));
var CommonFrontendContribution = /** @class */ (function () {
    function CommonFrontendContribution(shell, messageService) {
        this.shell = shell;
        this.messageService = messageService;
    }
    CommonFrontendContribution.prototype.registerMenus = function (registry) {
        registry.registerSubmenu(CommonMenus.FILE, 'File');
        registry.registerSubmenu(CommonMenus.EDIT, 'Edit');
        registry.registerSubmenu(CommonMenus.VIEW, 'View');
        registry.registerSubmenu(CommonMenus.HELP, 'Help');
        registry.registerMenuAction(CommonMenus.FILE_SAVE, {
            commandId: CommonCommands.SAVE.id
        });
        registry.registerMenuAction(CommonMenus.FILE_SAVE, {
            commandId: CommonCommands.SAVE_ALL.id
        });
        registry.registerMenuAction(CommonMenus.EDIT_UNDO, {
            commandId: CommonCommands.UNDO.id,
            order: '0'
        });
        registry.registerMenuAction(CommonMenus.EDIT_UNDO, {
            commandId: CommonCommands.REDO.id,
            order: '1'
        });
        registry.registerMenuAction(CommonMenus.EDIT_FIND, {
            commandId: CommonCommands.FIND.id,
            order: '0'
        });
        registry.registerMenuAction(CommonMenus.EDIT_FIND, {
            commandId: CommonCommands.REPLACE.id,
            order: '1'
        });
        registry.registerMenuAction(CommonMenus.EDIT_CLIPBOARD, {
            commandId: CommonCommands.CUT.id,
            order: '0'
        });
        registry.registerMenuAction(CommonMenus.EDIT_CLIPBOARD, {
            commandId: CommonCommands.COPY.id,
            order: '1'
        });
        registry.registerMenuAction(CommonMenus.EDIT_CLIPBOARD, {
            commandId: CommonCommands.PASTE.id,
            order: '2'
        });
        registry.registerMenuAction(CommonMenus.VIEW_LAYOUT, {
            commandId: CommonCommands.TOGGLE_BOTTOM_PANEL.id,
            order: '0'
        });
        registry.registerMenuAction(CommonMenus.VIEW_LAYOUT, {
            commandId: CommonCommands.COLLAPSE_ALL_PANELS.id,
            order: '1'
        });
        registry.registerMenuAction(tab_bars_1.SHELL_TABBAR_CONTEXT_MENU, {
            commandId: CommonCommands.CLOSE_TAB.id,
            label: 'Close',
            order: '0'
        });
        registry.registerMenuAction(tab_bars_1.SHELL_TABBAR_CONTEXT_MENU, {
            commandId: CommonCommands.CLOSE_OTHER_TABS.id,
            label: 'Close Others',
            order: '1'
        });
        registry.registerMenuAction(tab_bars_1.SHELL_TABBAR_CONTEXT_MENU, {
            commandId: CommonCommands.CLOSE_RIGHT_TABS.id,
            label: 'Close to the Right',
            order: '2'
        });
        registry.registerMenuAction(tab_bars_1.SHELL_TABBAR_CONTEXT_MENU, {
            commandId: CommonCommands.CLOSE_ALL_TABS.id,
            label: 'Close All',
            order: '3'
        });
        registry.registerMenuAction(tab_bars_1.SHELL_TABBAR_CONTEXT_MENU, {
            commandId: CommonCommands.COLLAPSE_PANEL.id,
            label: 'Collapse',
            order: '4'
        });
    };
    CommonFrontendContribution.prototype.registerCommands = function (commandRegistry) {
        var _this = this;
        commandRegistry.registerCommand(CommonCommands.CUT, {
            execute: function () {
                if (exports.supportCut) {
                    document.execCommand('cut');
                }
                else {
                    _this.messageService.warn("Please use the browser's cut command or shortcut.");
                }
            }
        });
        commandRegistry.registerCommand(CommonCommands.COPY, {
            execute: function () {
                if (exports.supportCopy) {
                    document.execCommand('copy');
                }
                else {
                    _this.messageService.warn("Please use the browser's copy command or shortcut.");
                }
            }
        });
        commandRegistry.registerCommand(CommonCommands.PASTE, {
            execute: function () {
                if (exports.supportPaste) {
                    document.execCommand('paste');
                }
                else {
                    _this.messageService.warn("Please use the browser's paste command or shortcut.");
                }
            }
        });
        commandRegistry.registerCommand(CommonCommands.UNDO);
        commandRegistry.registerCommand(CommonCommands.REDO);
        commandRegistry.registerCommand(CommonCommands.FIND);
        commandRegistry.registerCommand(CommonCommands.REPLACE);
        commandRegistry.registerCommand(CommonCommands.NEXT_TAB, {
            isEnabled: function () { return _this.shell.currentTabBar !== undefined; },
            execute: function () { return _this.shell.activateNextTab(); }
        });
        commandRegistry.registerCommand(CommonCommands.PREVIOUS_TAB, {
            isEnabled: function () { return _this.shell.currentTabBar !== undefined; },
            execute: function () { return _this.shell.activatePreviousTab(); }
        });
        commandRegistry.registerCommand(CommonCommands.CLOSE_TAB, {
            isEnabled: function () { return _this.shell.currentTabBar !== undefined; },
            execute: function () {
                var tabBar = _this.shell.currentTabBar;
                var currentTitle = tabBar.currentTitle;
                _this.shell.closeTabs(tabBar, function (title, index) { return title === currentTitle; });
            }
        });
        commandRegistry.registerCommand(CommonCommands.CLOSE_OTHER_TABS, {
            isEnabled: function () {
                var tabBar = _this.shell.currentTabBar;
                if (tabBar) {
                    return tabBar.titles.length > 1;
                }
                return false;
            },
            execute: function () {
                var tabBar = _this.shell.currentTabBar;
                var currentTitle = tabBar.currentTitle;
                _this.shell.closeTabs(_this.shell.currentTabArea, function (title, index) { return title !== currentTitle; });
            }
        });
        commandRegistry.registerCommand(CommonCommands.CLOSE_RIGHT_TABS, {
            isEnabled: function () {
                var tabBar = _this.shell.currentTabBar;
                if (tabBar) {
                    return tabBar.currentIndex < tabBar.titles.length - 1;
                }
                return false;
            },
            isVisible: function () {
                var area = _this.shell.currentTabArea;
                return area !== 'left' && area !== 'right';
            },
            execute: function () {
                var tabBar = _this.shell.currentTabBar;
                var currentIndex = tabBar.currentIndex;
                _this.shell.closeTabs(tabBar, function (title, index) { return index > currentIndex; });
            }
        });
        commandRegistry.registerCommand(CommonCommands.CLOSE_ALL_TABS, {
            isEnabled: function () { return _this.shell.currentTabBar !== undefined; },
            execute: function () { return _this.shell.closeTabs(_this.shell.currentTabArea); }
        });
        commandRegistry.registerCommand(CommonCommands.COLLAPSE_PANEL, {
            isEnabled: function () { return application_shell_1.ApplicationShell.isSideArea(_this.shell.currentTabArea); },
            isVisible: function () { return application_shell_1.ApplicationShell.isSideArea(_this.shell.currentTabArea); },
            execute: function () {
                var currentArea = _this.shell.currentTabArea;
                if (application_shell_1.ApplicationShell.isSideArea(currentArea)) {
                    _this.shell.collapsePanel(currentArea);
                }
            }
        });
        commandRegistry.registerCommand(CommonCommands.COLLAPSE_ALL_PANELS, {
            execute: function () {
                _this.shell.collapsePanel('left');
                _this.shell.collapsePanel('right');
                _this.shell.collapsePanel('bottom');
            }
        });
        commandRegistry.registerCommand(CommonCommands.TOGGLE_BOTTOM_PANEL, {
            isEnabled: function () { return _this.shell.getWidgets('bottom').length > 0; },
            execute: function () {
                if (_this.shell.isExpanded('bottom')) {
                    _this.shell.collapsePanel('bottom');
                }
                else {
                    _this.shell.expandPanel('bottom');
                }
            }
        });
        commandRegistry.registerCommand(CommonCommands.SAVE, {
            execute: function () { return _this.shell.save(); }
        });
        commandRegistry.registerCommand(CommonCommands.SAVE_ALL, {
            execute: function () { return _this.shell.saveAll(); }
        });
        commandRegistry.registerCommand(CommonCommands.QUIT, {
            execute: function () {
                /* FIXME implement QUIT of innermost command.  */
            }
        });
    };
    CommonFrontendContribution.prototype.registerKeybindings = function (registry) {
        if (exports.supportCut) {
            registry.registerKeybinding({
                command: CommonCommands.CUT.id,
                keybinding: "ctrlcmd+x"
            });
        }
        if (exports.supportCopy) {
            registry.registerKeybinding({
                command: CommonCommands.COPY.id,
                keybinding: "ctrlcmd+c"
            });
        }
        if (exports.supportPaste) {
            registry.registerKeybinding({
                command: CommonCommands.PASTE.id,
                keybinding: "ctrlcmd+v"
            });
        }
        registry.registerKeybindings({
            command: CommonCommands.UNDO.id,
            keybinding: "ctrlcmd+z"
        }, {
            command: CommonCommands.REDO.id,
            keybinding: "ctrlcmd+shift+z"
        }, {
            command: CommonCommands.FIND.id,
            keybinding: "ctrlcmd+f"
        }, {
            command: CommonCommands.REPLACE.id,
            keybinding: "ctrlcmd+alt+f"
        }, {
            command: CommonCommands.NEXT_TAB.id,
            keybinding: "ctrlcmd+tab"
        }, {
            command: CommonCommands.PREVIOUS_TAB.id,
            keybinding: "ctrlcmd+shift+tab"
        }, {
            command: CommonCommands.CLOSE_TAB.id,
            keybinding: "alt+w"
        }, {
            command: CommonCommands.CLOSE_OTHER_TABS.id,
            keybinding: "ctrlcmd+alt+t"
        }, {
            command: CommonCommands.CLOSE_ALL_TABS.id,
            keybinding: "alt+shift+w"
        }, {
            command: CommonCommands.COLLAPSE_PANEL.id,
            keybinding: "alt+c"
        }, {
            command: CommonCommands.TOGGLE_BOTTOM_PANEL.id,
            keybinding: "ctrlcmd+j",
        }, {
            command: CommonCommands.COLLAPSE_ALL_PANELS.id,
            keybinding: "alt+shift+c",
        }, {
            command: CommonCommands.SAVE.id,
            keybinding: "ctrlcmd+s"
        }, {
            command: CommonCommands.SAVE_ALL.id,
            keybinding: "ctrlcmd+alt+s"
        }, {
            command: CommonCommands.QUIT.id,
            keybinding: "ctrlcmd+q"
        });
    };
    CommonFrontendContribution = __decorate([
        inversify_1.injectable(),
        __param(0, inversify_1.inject(application_shell_1.ApplicationShell)),
        __param(1, inversify_1.inject(message_service_1.MessageService)),
        __metadata("design:paramtypes", [application_shell_1.ApplicationShell,
            message_service_1.MessageService])
    ], CommonFrontendContribution);
    return CommonFrontendContribution;
}());
exports.CommonFrontendContribution = CommonFrontendContribution;
