// @theia/workspace/lib/browser/workspace-uri-contribution.js
"use strict";
/*
 * Copyright (C) 2017 TypeFox and others.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var label_provider_1 = require("@theia/core/lib/browser/label-provider");
var uri_1 = require("@theia/core/lib/common/uri");
var inversify_1 = require("inversify");
var workspace_service_1 = require("./workspace-service");
var common_1 = require("@theia/filesystem/lib/common");
var WorkspaceUriLabelProviderContribution = /** @class */ (function (_super) {
    __extends(WorkspaceUriLabelProviderContribution, _super);
    function WorkspaceUriLabelProviderContribution(wsService, fileSystem) {
        var _this = _super.call(this) || this;
        _this.fileSystem = fileSystem;
        wsService.root.then(function (root) {
            if (root) {
                _this.wsRoot = root.uri;
            }
        });
        return _this;
    }
    WorkspaceUriLabelProviderContribution.prototype.canHandle = function (element) {
        if ((element instanceof uri_1.default && element.scheme === 'file' || common_1.FileStat.is(element))) {
            return 10;
        }
        return 0;
    };
    WorkspaceUriLabelProviderContribution.prototype.getUri = function (element) {
        if (common_1.FileStat.is(element)) {
            return new uri_1.default(element.uri);
        }
        return new uri_1.default(element.toString());
    };
    WorkspaceUriLabelProviderContribution.prototype.getStat = function (element) {
        if (common_1.FileStat.is(element)) {
            return element;
        }
        return this.fileSystem.getFileStat(element.toString());
    };
    WorkspaceUriLabelProviderContribution.prototype.getIcon = function (element) {
        return __awaiter(this, void 0, void 0, function () {
            var uri, icon, stat, err_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (common_1.FileStat.is(element) && element.isDirectory) {
                            return [2 /*return*/, 'fa fa-folder'];
                        }
                        uri = this.getUri(element);
                        icon = _super.prototype.getFileIcon.call(this, uri);
                        if (!!icon) return [3 /*break*/, 4];
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this.getStat(element)];
                    case 2:
                        stat = _a.sent();
                        if (stat.isDirectory) {
                            return [2 /*return*/, 'fa fa-folder'];
                        }
                        else {
                            return [2 /*return*/, 'fa fa-file'];
                        }
                        return [3 /*break*/, 4];
                    case 3:
                        err_1 = _a.sent();
                        return [2 /*return*/, 'fa fa-file'];
                    case 4: return [2 /*return*/, icon];
                }
            });
        });
    };
    WorkspaceUriLabelProviderContribution.prototype.getName = function (element) {
        return _super.prototype.getName.call(this, this.getUri(element));
    };
    /**
     * trims the workspace root from a file uri, if it is a child.
     */
    WorkspaceUriLabelProviderContribution.prototype.getLongName = function (element) {
        var uri = this.getUri(element);
        var uriStr = uri.toString();
        if (!this.wsRoot || !uriStr.startsWith(this.wsRoot)) {
            return _super.prototype.getLongName.call(this, uri);
        }
        var short = uri.toString().substr(this.wsRoot.length);
        if (short[0] === '/') {
            return short.substr(1);
        }
        return short;
    };
    WorkspaceUriLabelProviderContribution = __decorate([
        inversify_1.injectable(),
        __param(0, inversify_1.inject(workspace_service_1.WorkspaceService)),
        __param(1, inversify_1.inject(common_1.FileSystem)),
        __metadata("design:paramtypes", [workspace_service_1.WorkspaceService, Object])
    ], WorkspaceUriLabelProviderContribution);
    return WorkspaceUriLabelProviderContribution;
}(label_provider_1.DefaultUriLabelProviderContribution));
exports.WorkspaceUriLabelProviderContribution = WorkspaceUriLabelProviderContribution;
