// @theia/git/lib/browser/git-navigable-list-widget.js
"use strict";
/*
 * Copyright (C) 2018 TypeFox and others.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var browser_1 = require("@theia/core/lib/browser");
var common_1 = require("../common");
var uri_1 = require("@theia/core/lib/common/uri");
var git_repository_provider_1 = require("./git-repository-provider");
var label_provider_1 = require("@theia/core/lib/browser/label-provider");
var domutils_1 = require("@phosphor/domutils");
var inversify_1 = require("inversify");
var GitNavigableListWidget = /** @class */ (function (_super) {
    __extends(GitNavigableListWidget, _super);
    function GitNavigableListWidget() {
        var _this = _super.call(this) || this;
        _this.node.tabIndex = 0;
        return _this;
    }
    GitNavigableListWidget.prototype.onActivateRequest = function (msg) {
        _super.prototype.onActivateRequest.call(this, msg);
        this.node.focus();
    };
    Object.defineProperty(GitNavigableListWidget.prototype, "scrollContainer", {
        get: function () {
            return this._scrollContainer;
        },
        set: function (id) {
            this._scrollContainer = id + Date.now();
        },
        enumerable: true,
        configurable: true
    });
    GitNavigableListWidget.prototype.onUpdateRequest = function (msg) {
        var _this = this;
        _super.prototype.onUpdateRequest.call(this, msg);
        (function () { return __awaiter(_this, void 0, void 0, function () {
            var selected, scrollArea;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        selected = this.node.getElementsByClassName(browser_1.SELECTED_CLASS)[0];
                        return [4 /*yield*/, this.getScrollContainer()];
                    case 1:
                        scrollArea = _a.sent();
                        if (selected && scrollArea) {
                            domutils_1.ElementExt.scrollIntoViewIfNeeded(scrollArea, selected);
                        }
                        return [2 /*return*/];
                }
            });
        }); })();
    };
    GitNavigableListWidget.prototype.getStatusCaption = function (status, staged) {
        return common_1.GitFileStatus.toString(status, staged);
    };
    GitNavigableListWidget.prototype.getAbbreviatedStatusCaption = function (status, staged) {
        return common_1.GitFileStatus.toAbbreviation(status, staged);
    };
    GitNavigableListWidget.prototype.relativePath = function (uri) {
        var parsedUri = typeof uri === 'string' ? new uri_1.default(uri) : uri;
        var repo = this.repositoryProvider.selectedRepository;
        if (repo) {
            return common_1.Repository.relativePath(repo, parsedUri).toString();
        }
        else {
            return this.labelProvider.getLongName(parsedUri);
        }
    };
    GitNavigableListWidget.prototype.computeCaption = function (fileChange) {
        var result = this.relativePath(fileChange.uri) + " - " + this.getStatusCaption(fileChange.status, true);
        if (fileChange.oldUri) {
            result = this.relativePath(fileChange.oldUri) + " -> " + result;
        }
        return result;
    };
    GitNavigableListWidget.prototype.onAfterAttach = function (msg) {
        var _this = this;
        _super.prototype.onAfterAttach.call(this, msg);
        this.addKeyListener(this.node, browser_1.Key.ARROW_LEFT, function () { return _this.navigateLeft(); });
        this.addKeyListener(this.node, browser_1.Key.ARROW_RIGHT, function () { return _this.navigateRight(); });
        this.addKeyListener(this.node, browser_1.Key.ARROW_UP, function () { return _this.navigateUp(); });
        this.addKeyListener(this.node, browser_1.Key.ARROW_DOWN, function () { return _this.navigateDown(); });
        this.addKeyListener(this.node, browser_1.Key.ENTER, function () { return _this.handleListEnter(); });
    };
    GitNavigableListWidget.prototype.navigateLeft = function () {
        this.selectPreviousNode();
    };
    GitNavigableListWidget.prototype.navigateRight = function () {
        this.selectNextNode();
    };
    GitNavigableListWidget.prototype.navigateUp = function () {
        this.selectPreviousNode();
    };
    GitNavigableListWidget.prototype.navigateDown = function () {
        this.selectNextNode();
    };
    GitNavigableListWidget.prototype.handleListEnter = function () {
    };
    GitNavigableListWidget.prototype.getSelected = function () {
        return this.gitNodes ? this.gitNodes.find(function (c) { return c.selected || false; }) : undefined;
    };
    GitNavigableListWidget.prototype.selectNode = function (node) {
        var n = this.getSelected();
        if (n) {
            n.selected = false;
        }
        node.selected = true;
        this.update();
    };
    GitNavigableListWidget.prototype.selectNextNode = function () {
        var idx = this.indexOfSelected;
        if (idx >= 0 && idx < this.gitNodes.length - 1) {
            this.selectNode(this.gitNodes[idx + 1]);
        }
        else if (this.gitNodes.length > 0 && idx === -1) {
            this.selectNode(this.gitNodes[0]);
        }
    };
    GitNavigableListWidget.prototype.selectPreviousNode = function () {
        var idx = this.indexOfSelected;
        if (idx > 0) {
            this.selectNode(this.gitNodes[idx - 1]);
        }
    };
    Object.defineProperty(GitNavigableListWidget.prototype, "indexOfSelected", {
        get: function () {
            if (this.gitNodes && this.gitNodes.length > 0) {
                return this.gitNodes.findIndex(function (c) { return c.selected || false; });
            }
            return -1;
        },
        enumerable: true,
        configurable: true
    });
    __decorate([
        inversify_1.inject(git_repository_provider_1.GitRepositoryProvider),
        __metadata("design:type", git_repository_provider_1.GitRepositoryProvider)
    ], GitNavigableListWidget.prototype, "repositoryProvider", void 0);
    __decorate([
        inversify_1.inject(label_provider_1.LabelProvider),
        __metadata("design:type", label_provider_1.LabelProvider)
    ], GitNavigableListWidget.prototype, "labelProvider", void 0);
    GitNavigableListWidget = __decorate([
        inversify_1.injectable(),
        __metadata("design:paramtypes", [])
    ], GitNavigableListWidget);
    return GitNavigableListWidget;
}(browser_1.VirtualWidget));
exports.GitNavigableListWidget = GitNavigableListWidget;
