// @theia/navigator/lib/browser/navigator-widget.js
"use strict";
/*
 * Copyright (C) 2017 TypeFox and others.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var lib_1 = require("@phosphor/virtualdom/lib");
var uri_1 = require("@theia/core/lib/common/uri");
var common_1 = require("@theia/core/lib/common");
var browser_1 = require("@theia/core/lib/browser");
var browser_2 = require("@theia/filesystem/lib/browser");
var browser_3 = require("@theia/workspace/lib/browser");
var navigator_model_1 = require("./navigator-model");
exports.FILE_NAVIGATOR_ID = 'files';
exports.LABEL = 'Files';
exports.CLASS = 'theia-Files';
var FileNavigatorWidget = /** @class */ (function (_super) {
    __extends(FileNavigatorWidget, _super);
    function FileNavigatorWidget(props, model, contextMenuRenderer, commandService, selectionService, workspaceService, labelProvider) {
        var _this = _super.call(this, props, model, contextMenuRenderer) || this;
        _this.props = props;
        _this.model = model;
        _this.commandService = commandService;
        _this.selectionService = selectionService;
        _this.workspaceService = workspaceService;
        _this.labelProvider = labelProvider;
        _this.id = exports.FILE_NAVIGATOR_ID;
        _this.title.label = exports.LABEL;
        _this.addClass(exports.CLASS);
        _this.initialize();
        return _this;
    }
    FileNavigatorWidget.prototype.initialize = function () {
        var _this = this;
        this.model.onSelectionChanged(function (selection) {
            return _this.selectionService.selection = selection;
        });
        this.workspaceService.root.then(function (resolvedRoot) { return __awaiter(_this, void 0, void 0, function () {
            var uri, label, icon;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!resolvedRoot) return [3 /*break*/, 2];
                        uri = new uri_1.default(resolvedRoot.uri);
                        label = this.labelProvider.getName(uri);
                        return [4 /*yield*/, this.labelProvider.getIcon(resolvedRoot)];
                    case 1:
                        icon = _a.sent();
                        this.model.root = browser_2.DirNode.createRoot(resolvedRoot, label, icon);
                        return [3 /*break*/, 3];
                    case 2:
                        this.update();
                        _a.label = 3;
                    case 3: return [2 /*return*/];
                }
            });
        }); });
    };
    FileNavigatorWidget.prototype.deflateForStorage = function (node) {
        var copy = Object.assign({}, node);
        if (copy.uri) {
            copy.uri = copy.uri.toString();
        }
        return _super.prototype.deflateForStorage.call(this, copy);
    };
    FileNavigatorWidget.prototype.inflateFromStorage = function (node, parent) {
        if (node.uri) {
            node.uri = new uri_1.default(node.uri);
        }
        return _super.prototype.inflateFromStorage.call(this, node, parent);
    };
    FileNavigatorWidget.prototype.renderTree = function (model) {
        return _super.prototype.renderTree.call(this, model) || this.renderOpenWorkspaceDiv();
    };
    FileNavigatorWidget.prototype.onAfterAttach = function (msg) {
        var _this = this;
        _super.prototype.onAfterAttach.call(this, msg);
        this.addClipboardListener(this.node, 'copy', function (e) { return _this.handleCopy(e); });
        this.addClipboardListener(this.node, 'paste', function (e) { return _this.handlePaste(e); });
    };
    FileNavigatorWidget.prototype.handleCopy = function (event) {
        var node = this.model.selectedFileStatNode;
        if (!node) {
            return;
        }
        var uri = node.uri.toString();
        event.clipboardData.setData('text/plain', uri);
        event.preventDefault();
    };
    FileNavigatorWidget.prototype.handlePaste = function (event) {
        var raw = event.clipboardData.getData('text/plain');
        if (!raw) {
            return;
        }
        var uri = new uri_1.default(raw);
        if (this.model.copy(uri)) {
            event.preventDefault();
        }
    };
    /**
     * Instead of rendering the file resources form the workspace, we render a placeholder
     * button when the workspace root is not yet set.
     */
    FileNavigatorWidget.prototype.renderOpenWorkspaceDiv = function () {
        var _this = this;
        var button = lib_1.h.button({
            className: 'open-workspace-button',
            title: 'Select a directory as your workspace root',
            onclick: function (e) { return _this.commandService.executeCommand(browser_3.WorkspaceCommands.OPEN.id); }
        }, 'Open Workspace');
        var buttonContainer = lib_1.h.div({ className: 'open-workspace-button-container' }, button);
        return lib_1.h.div({ className: 'theia-navigator-container' }, 'You have not yet opened a workspace.', buttonContainer);
    };
    FileNavigatorWidget = __decorate([
        inversify_1.injectable(),
        __param(0, inversify_1.inject(browser_1.TreeProps)),
        __param(1, inversify_1.inject(navigator_model_1.FileNavigatorModel)),
        __param(2, inversify_1.inject(browser_1.ContextMenuRenderer)),
        __param(3, inversify_1.inject(common_1.CommandService)),
        __param(4, inversify_1.inject(common_1.SelectionService)),
        __param(5, inversify_1.inject(browser_3.WorkspaceService)),
        __param(6, inversify_1.inject(browser_1.LabelProvider)),
        __metadata("design:paramtypes", [Object, navigator_model_1.FileNavigatorModel, Object, Object, common_1.SelectionService,
            browser_3.WorkspaceService,
            browser_1.LabelProvider])
    ], FileNavigatorWidget);
    return FileNavigatorWidget;
}(browser_2.FileTreeWidget));
exports.FileNavigatorWidget = FileNavigatorWidget;
