// @theia/git/lib/node/git-backend-module.js
"use strict";
/*
 * Copyright (C) 2017 TypeFox and others.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 */
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
var cluster = require("cluster");
var inversify_1 = require("inversify");
var git_1 = require("../common/git");
var git_watcher_1 = require("../common/git-watcher");
var dugite_git_1 = require("./dugite-git");
var dugite_git_watcher_1 = require("./dugite-git-watcher");
var common_1 = require("@theia/core/lib/common");
var git_repository_manager_1 = require("./git-repository-manager");
var git_repository_watcher_1 = require("./git-repository-watcher");
var git_locator_protocol_1 = require("./git-locator/git-locator-protocol");
var git_locator_client_1 = require("./git-locator/git-locator-client");
var git_locator_impl_1 = require("./git-locator/git-locator-impl");
var GitBindingOptions;
(function (GitBindingOptions) {
    GitBindingOptions.Default = {
        bindManager: function (binding) {
            return binding.to(git_repository_manager_1.GitRepositoryManager).inSingletonScope();
        }
    };
})(GitBindingOptions = exports.GitBindingOptions || (exports.GitBindingOptions = {}));
function bindGit(bind, bindingOptions) {
    if (bindingOptions === void 0) { bindingOptions = GitBindingOptions.Default; }
    bindingOptions.bindManager(bind(git_repository_manager_1.GitRepositoryManager));
    bind(git_repository_watcher_1.GitRepositoryWatcherFactory).toFactory(function (ctx) { return function (options) {
        var child = new inversify_1.Container({ defaultScope: 'Singleton' });
        child.parent = ctx.container;
        child.bind(git_repository_watcher_1.GitRepositoryWatcher).toSelf();
        child.bind(git_repository_watcher_1.GitRepositoryWatcherOptions).toConstantValue(options);
        return child.get(git_repository_watcher_1.GitRepositoryWatcher);
    }; });
    if (cluster.isMaster) {
        bind(git_locator_protocol_1.GitLocator).toDynamicValue(function (ctx) {
            var logger = ctx.container.get(common_1.ILogger);
            return new git_locator_impl_1.GitLocatorImpl({
                info: function (message) {
                    var args = [];
                    for (var _i = 1; _i < arguments.length; _i++) {
                        args[_i - 1] = arguments[_i];
                    }
                    return logger.info.apply(logger, __spread([message], args));
                },
                error: function (message) {
                    var args = [];
                    for (var _i = 1; _i < arguments.length; _i++) {
                        args[_i - 1] = arguments[_i];
                    }
                    return logger.error.apply(logger, __spread([message], args));
                }
            });
        });
    }
    else {
        bind(git_locator_protocol_1.GitLocator).to(git_locator_client_1.GitLocatorClient);
    }
    bind(dugite_git_1.DugiteGit).toSelf();
    bind(dugite_git_1.OutputParser).toSelf();
    bind(dugite_git_1.NameStatusParser).toSelf();
    bind(dugite_git_1.CommitDetailsParser).toSelf();
    bind(git_1.Git).toDynamicValue(function (ctx) { return ctx.container.get(dugite_git_1.DugiteGit); });
}
exports.bindGit = bindGit;
function bindRepositoryWatcher(bind) {
    bind(dugite_git_watcher_1.DugiteGitWatcherServer).toSelf();
    bind(git_watcher_1.GitWatcherServer).toDynamicValue(function (context) { return context.container.get(dugite_git_watcher_1.DugiteGitWatcherServer); });
}
exports.bindRepositoryWatcher = bindRepositoryWatcher;
exports.default = new inversify_1.ContainerModule(function (bind) {
    bindGit(bind);
    bind(common_1.ConnectionHandler).toDynamicValue(function (context) {
        return new common_1.JsonRpcConnectionHandler(git_1.GitPath, function (client) {
            var server = context.container.get(git_1.Git);
            client.onDidCloseConnection(function () { return server.dispose(); });
            return server;
        });
    }).inSingletonScope();
    bindRepositoryWatcher(bind);
    bind(common_1.ConnectionHandler).toDynamicValue(function (context) {
        return new common_1.JsonRpcConnectionHandler(git_watcher_1.GitWatcherPath, function (client) {
            var server = context.container.get(git_watcher_1.GitWatcherServer);
            server.setClient(client);
            client.onDidCloseConnection(function () { return server.dispose(); });
            return server;
        });
    }).inSingletonScope();
});
