// @theia/merge-conflicts/lib/browser/merge-conflicts-parser.js
"use strict";
/*
 * Copyright (C) 2018 TypeFox and others.
 *
 * Licensed under the Apache License, Version 2.0 (the 'License'); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var MergeConflictsParser = /** @class */ (function () {
    function MergeConflictsParser() {
        this.init();
    }
    MergeConflictsParser_1 = MergeConflictsParser;
    MergeConflictsParser.prototype.parse = function (text) {
        var context = new MergeConflictsParser_1.Context();
        this.parser.reset(context);
        var lines = text.split(/\r?\n|\r/);
        for (var line = 0; line < lines.length; line++) {
            this.parser.read({ line: line, text: lines[line] });
        }
        return context.all;
    };
    MergeConflictsParser.prototype.init = function () {
        var _this = this;
        var parser = this.parser = new MergeConflictsParser_1.StateMachine(new MergeConflictsParser_1.Context());
        // states
        var start = parser.addState('start');
        var currentMarker = parser.addState('current-marker');
        var currentContent = parser.addState('current-content');
        var baseMarker = parser.addState('base-marker');
        var baseContent = parser.addState('base-content');
        var separator = parser.addState('separator');
        var incomingContent = parser.addState('incoming-content');
        var incomingMarker = parser.addState('incoming-marker');
        var push = parser.addState('push');
        // conditions / triggers
        var startsWithLt = function (input) { return input.text.startsWith('<<<<<<<'); };
        var startsWithEq = function (input) { return input.text.startsWith('======='); };
        var startsWithGt = function (input) { return input.text.startsWith('>>>>>>>'); };
        var startsWithPp = function (input) { return input.text.startsWith('|||||||'); };
        var any = function () { return true; };
        // transitions
        start.to(currentMarker, startsWithLt);
        start.to(start, any);
        currentMarker.to(currentMarker, startsWithLt);
        currentMarker.to(separator, startsWithEq);
        currentMarker.to(currentContent, any);
        currentContent.to(currentMarker, startsWithLt);
        currentContent.to(separator, startsWithEq);
        currentContent.to(baseMarker, startsWithPp);
        currentContent.to(start, startsWithGt);
        currentContent.to(currentContent, any);
        baseMarker.to(currentMarker, startsWithLt);
        baseMarker.to(baseMarker, startsWithPp);
        baseMarker.to(separator, startsWithEq);
        baseMarker.to(baseContent, any);
        baseContent.to(currentMarker, startsWithLt);
        baseContent.to(separator, startsWithEq);
        baseContent.to(baseMarker, startsWithPp);
        baseContent.to(baseContent, any);
        separator.to(currentMarker, startsWithLt);
        separator.to(start, startsWithEq);
        separator.to(incomingContent, any);
        incomingContent.to(start, startsWithEq);
        incomingContent.to(currentMarker, startsWithLt);
        incomingContent.to(incomingMarker, startsWithGt);
        incomingContent.to(incomingContent, any);
        incomingMarker.to(push);
        push.to(start);
        // actions
        currentMarker.onEnter = function (input, ctx) {
            ctx.new = new MergeConflictsParser_1.Context().new;
            ctx.new.current.marker = _this.lineToRange(input);
        };
        currentContent.onEnter = function (input, ctx) {
            var current = ctx.new.current;
            current.content = _this.addLineToRange(input, current.content);
        };
        baseMarker.onEnter = function (input, ctx) {
            ctx.new.bases.push({ marker: _this.lineToRange(input) });
        };
        baseContent.onEnter = function (input, ctx) {
            var base = ctx.new.bases.slice(-1)[0];
            base.content = _this.addLineToRange(input, base.content);
        };
        incomingContent.onEnter = function (input, ctx) {
            var incoming = ctx.new.incoming;
            incoming.content = _this.addLineToRange(input, incoming.content);
        };
        incomingMarker.onEnter = function (input, ctx) {
            var markerRange = _this.lineToRange(input);
            ctx.new.incoming.marker = markerRange;
            ctx.new.total = {
                start: ctx.new.current.marker.start,
                end: markerRange.end
            };
        };
        push.onEnter = function (input, ctx) {
            ctx.all.push(ctx.new);
        };
    };
    MergeConflictsParser.prototype.lineToRange = function (input) {
        return {
            start: { line: input.line, character: 0 },
            end: { line: input.line, character: input.text.length },
        };
    };
    MergeConflictsParser.prototype.addLineToRange = function (input, range) {
        if (!range) {
            return this.lineToRange(input);
        }
        range.end = { line: input.line, character: input.text.length };
        return range;
    };
    MergeConflictsParser = MergeConflictsParser_1 = __decorate([
        inversify_1.injectable(),
        __metadata("design:paramtypes", [])
    ], MergeConflictsParser);
    return MergeConflictsParser;
    var MergeConflictsParser_1;
}());
exports.MergeConflictsParser = MergeConflictsParser;
(function (MergeConflictsParser) {
    var Context = /** @class */ (function () {
        function Context() {
            this.new = {
                current: {},
                incoming: {},
                bases: []
            };
            this.all = [];
        }
        return Context;
    }());
    MergeConflictsParser.Context = Context;
    var StateMachine = /** @class */ (function () {
        function StateMachine(context) {
            this.context = context;
            this.states = [];
        }
        StateMachine.prototype.reset = function (context) {
            this.current = this.states[0];
            this.context = context;
        };
        StateMachine.prototype.read = function (input) {
            var next = this.current.findNext(input, this.context);
            while (next) {
                if (next.onEnter) {
                    next.onEnter(input, this.context);
                }
                if (next.immediateNext) {
                    this.current = next;
                    next = next.immediateNext;
                }
                else {
                    break;
                }
            }
            if (!next) {
                throw new Error("Missing transition from (" + this.current.id + ") for input: L." + input.line + " > " + input.text);
            }
            this.current = next;
        };
        StateMachine.prototype.addState = function (id) {
            var state = new State(id);
            this.states.push(state);
            if (!this.current) {
                this.current = state;
            }
            return state;
        };
        return StateMachine;
    }());
    MergeConflictsParser.StateMachine = StateMachine;
    var State = /** @class */ (function () {
        function State(id) {
            this.id = id;
            this.conditionalNext = [];
        }
        State.prototype.findNext = function (input, context) {
            var match = this.conditionalNext.find(function (transition) { return transition.condition(input, context); });
            return match ? match.to : undefined;
        };
        State.prototype.to = function (next, condition) {
            if (condition) {
                this.immediateNext = undefined;
                this.conditionalNext.push({ to: next, condition: condition });
            }
            else {
                this.immediateNext = next;
            }
        };
        return State;
    }());
    MergeConflictsParser.State = State;
})(MergeConflictsParser = exports.MergeConflictsParser || (exports.MergeConflictsParser = {}));
exports.MergeConflictsParser = MergeConflictsParser;
